File: packages/core-py/src/arky_core/settler.py
"""Settler execution per ARKY-SETTLERS-v1 sections 4 (pre-checks) and 5 (XR).

Validate an execution request (verb registered, args valid per the verb schema,
rail supported), then produce a signed Execution Receipt. Deterministic
idempotency (section 6.1) dedupes duplicate requests.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from enum import Enum

from .canonicalize import canonicalize
from .cid import cid_from_canonical, multihash_mb
from .jsonvalue import JsonObject, Number, path
from .jws import sign_detached
from .kernel import is_registered_verb

__all__ = [
    "ExecRequest",
    "ExecStatus",
    "ExecuteResult",
    "IdempotencyStore",
    "args_hash",
    "derive_idempotency_key",
    "execute",
    "verb_required_args",
]

_REQUIRED_ARGS: dict[str, tuple[str, ...]] = {
    "arky:verb/pay@v1": ("to", "amount"),
    "arky:verb/refund@v1": ("payment_ref",),
    "arky:verb/slash@v1": ("subject", "amount"),
    "arky:verb/revoke@v1": ("subject",),
    "arky:verb/upgrade@v1": ("target", "version"),
    "arky:verb/signal@v1": ("channel",),
    "arky:verb/control@v1": ("action",),
}


def verb_required_args(verb: str) -> tuple[str, ...]:
    """Required argument fields for a core verb, per ``schemas/verbs/*.json``.

    The verb schemas are canonical here: the section 3.2 prose was reconciled to
    them after an audit found a three-way contradiction between the spec text,
    the schemas, and the S1 vectors.
    """
    return _REQUIRED_ARGS.get(verb, ())


class ExecStatus(Enum):
    """The outcome of an execution attempt."""

    SUCCESS = "SUCCESS"
    FAILED = "FAILED"

    def __str__(self) -> str:
        return self.value


@dataclass
class ExecRequest:
    """Inputs to an execution."""

    verb: str
    args: object
    rail: str = ""
    commitment_cid: str = ""
    request_id: str = ""
    idempotency_key: str = ""


@dataclass
class ExecuteResult:
    """The result of :func:`execute`."""

    status: ExecStatus
    errors: list[str] = field(default_factory=list)
    missing_fields: list[str] = field(default_factory=list)
    receipt: object | None = None  # the signed XR, only on success


#: In-memory idempotency store: key -> XR.
IdempotencyStore = dict[str, object]


def _rail_supported(rail: str) -> bool:
    """An explicit ``unknown:`` scheme is unsupported; an absent rail is fine."""
    return not rail.startswith("unknown:")


def _validate_amount(args: object) -> str:
    """Enforce section 3.2 on a present ``amount``.

    It MUST be ``{value: finite number > 0, unit: non-empty string}``. Returns
    the offending field, or "" when absent or valid.

    This is the difference between a Settler that authorizes a negative payment
    and one that does not: an audit found every stack approving
    ``{value: -1000}`` because they only checked that the key existed.
    """
    amount, found = path(args, "amount")
    if not found:
        return ""
    if not isinstance(amount, JsonObject):
        return "amount"
    value = amount.get("value")
    if not isinstance(value, Number):
        return "amount.value"
    try:
        number = float(str(value))
    except ValueError:
        return "amount.value"
    if math.isnan(number) or math.isinf(number) or number <= 0:
        return "amount.value"
    unit = amount.get("unit")
    if not isinstance(unit, str) or isinstance(unit, Number) or not unit:
        return "amount.unit"
    return ""


def args_hash(args: object) -> str:
    """multibase(multihash(sha2-256, JCS(args)))."""
    return multihash_mb(canonicalize(args).encode("utf-8"))


def derive_idempotency_key(
    commitment_cid: str, verb: str, rail: str, args: object, verb_index: int = 0
) -> str:
    """Derive an idempotency key per section 6.1 when the client omits one.

    Deterministic and JCS-based, so the same request derives the same key in
    every stack.
    """
    components = JsonObject()
    components.set("args_hash", args_hash(args))
    components.set("commitment_cid", commitment_cid)
    components.set("rail", rail)
    components.set("verb", verb)
    components.set("verb_index", Number(str(verb_index)))
    return multihash_mb(canonicalize(components).encode("utf-8"))


def execute(
    request: ExecRequest,
    seed: bytes,
    kid: str = "",
    ts: str = "",
    anchor_target: str = "",
    store: IdempotencyStore | None = None,
) -> ExecuteResult:
    """Validate and execute a request, returning a signed Execution Receipt.

    Pre-check order is verb, then args, then rail (section 4.2). There is no
    real rail, so the XR carries a mock locator and a pending anchor. Pass a
    ``store`` to enable idempotency: a repeat of the same key returns the
    identical cached receipt rather than executing twice.
    """
    if not is_registered_verb(request.verb):
        return ExecuteResult(ExecStatus.FAILED, errors=["settler.unknown_verb"])

    missing = [k for k in verb_required_args(request.verb) if not path(request.args, k)[1]]
    if missing:
        return ExecuteResult(
            ExecStatus.FAILED, errors=["settler.invalid_args"], missing_fields=missing
        )

    bad_field = _validate_amount(request.args)
    if bad_field:
        return ExecuteResult(
            ExecStatus.FAILED, errors=["settler.invalid_args"], missing_fields=[bad_field]
        )

    if not _rail_supported(request.rail):
        return ExecuteResult(ExecStatus.FAILED, errors=["settler.unsupported_rail"])

    idempotency_key = request.idempotency_key or derive_idempotency_key(
        request.commitment_cid, request.verb, request.rail, request.args, 0
    )

    if store is not None and idempotency_key in store:
        return ExecuteResult(ExecStatus.SUCCESS, receipt=store[idempotency_key])

    request_id = request.request_id or f"exec-{idempotency_key[:12]}"

    anchor = JsonObject()
    anchor.set("target", anchor_target)
    anchor.set("locator", f"batch-{idempotency_key[1:10]}")
    anchor.set("status", "pending")

    body = JsonObject()
    body.set("request_id", request_id)
    body.set("commitment_cid", request.commitment_cid)
    body.set("verb", request.verb)
    body.set("rail", request.rail)
    body.set("args_hash", args_hash(request.args))
    body.set("idempotency_key", idempotency_key)
    body.set("status", "success")
    body.set("locator", f"MOCK-{idempotency_key[1:18]}")
    body.set("anchors", [anchor])
    body.set("ts", ts)

    canonical = canonicalize(body)
    receipt = body.clone()
    receipt.set("cid", cid_from_canonical(canonical))
    receipt.set("sig", sign_detached(canonical.encode("utf-8"), seed, kid or None))

    if store is not None:
        store[idempotency_key] = receipt
    return ExecuteResult(ExecStatus.SUCCESS, receipt=receipt)

File: packages/core-go/settler.go
package arky

import (
	"crypto/ed25519"
	"math"
)

// VerbRequiredArgs returns the required argument fields for a core verb, per
// schemas/verbs/*.json (the canonical source; ARKY-SETTLERS-v1 section 3.2
// prose was reconciled to these).
func VerbRequiredArgs(verb string) []string {
	switch verb {
	case "arky:verb/pay@v1":
		return []string{"to", "amount"}
	case "arky:verb/refund@v1":
		return []string{"payment_ref"}
	case "arky:verb/slash@v1":
		return []string{"subject", "amount"}
	case "arky:verb/revoke@v1":
		return []string{"subject"}
	case "arky:verb/upgrade@v1":
		return []string{"target", "version"}
	case "arky:verb/signal@v1":
		return []string{"channel"}
	case "arky:verb/control@v1":
		return []string{"action"}
	}
	return nil
}

// ExecStatus is the outcome of an execution attempt.
type ExecStatus int

const (
	ExecFailed ExecStatus = iota
	ExecSuccess
)

func (e ExecStatus) String() string {
	if e == ExecSuccess {
		return "SUCCESS"
	}
	return "FAILED"
}

// ExecuteResult is the result of Execute.
type ExecuteResult struct {
	Status        ExecStatus
	Errors        []string
	MissingFields []string
	Receipt       Value // the signed XR, only on success
}

// ExecRequest describes an execution to attempt.
type ExecRequest struct {
	Verb           string
	Rail           string
	Args           Value
	CommitmentCid  string
	RequestID      string
	IdempotencyKey string
}

// IdempotencyStore maps an idempotency key to its XR.
type IdempotencyStore map[string]Value

// railSupported treats an explicit "unknown:" scheme as unsupported; an absent
// rail is fine.
func railSupported(rail string) bool {
	return len(rail) < 8 || rail[:8] != "unknown:"
}

// validateAmount enforces section 3.2: a present amount MUST be
// { value: finite number > 0, unit: non-empty string }. Returns the offending
// field, or "" when absent or valid.
//
// This check is the difference between a Settler that authorizes a negative
// payment and one that does not; an earlier audit found both stacks approving
// {value:-1000} because they only checked that the key existed.
func validateAmount(args Value) string {
	amount, ok := Path(args, "amount")
	if !ok {
		return ""
	}
	obj, ok := amount.(*Object)
	if !ok {
		return "amount"
	}
	v, ok := obj.Get("value")
	if !ok {
		return "amount.value"
	}
	n, ok := v.(Number)
	if !ok {
		return "amount.value"
	}
	f, err := n.Float()
	if err != nil || math.IsNaN(f) || math.IsInf(f, 0) || f <= 0 {
		return "amount.value"
	}
	u, ok := obj.Get("unit")
	if !ok {
		return "amount.unit"
	}
	s, ok := u.(string)
	if !ok || s == "" {
		return "amount.unit"
	}
	return ""
}

// ArgsHash returns multibase(multihash(sha2-256, JCS(args))).
func ArgsHash(args Value) (string, error) {
	canonical, err := Canonicalize(args)
	if err != nil {
		return "", err
	}
	return MultihashMb([]byte(canonical)), nil
}

// DeriveIdempotencyKey derives a key per ARKY-SETTLERS-v1 section 6.1, for
// clients that omit one. Deterministic and JCS-based, so the same request
// derives the same key in every stack.
func DeriveIdempotencyKey(commitmentCid, verb, rail string, args Value, verbIndex int) (string, error) {
	ah, err := ArgsHash(args)
	if err != nil {
		return "", err
	}
	components := NewObject()
	components.Set("args_hash", ah)
	components.Set("commitment_cid", commitmentCid)
	components.Set("rail", rail)
	components.Set("verb", verb)
	components.Set("verb_index", Number(itoa(verbIndex)))
	canonical, err := Canonicalize(components)
	if err != nil {
		return "", err
	}
	return MultihashMb([]byte(canonical)), nil
}

func truncate(s string, n int) string {
	if len(s) < n {
		return s
	}
	return s[:n]
}

// Execute validates and executes a request, returning a signed Execution
// Receipt. Pre-check order is verb, then args, then rail (section 4.2); there
// is no real rail, so the XR carries a mock locator and a pending anchor.
//
// Pass a non-nil store to enable idempotency (section 6.1): a repeat of the
// same key returns the identical cached receipt rather than executing twice.
func Execute(req ExecRequest, priv ed25519.PrivateKey, kid, ts, anchorTarget string, store IdempotencyStore) ExecuteResult {
	if !IsRegisteredVerb(req.Verb) {
		return ExecuteResult{Status: ExecFailed, Errors: []string{"settler.unknown_verb"}}
	}

	var missing []string
	for _, k := range VerbRequiredArgs(req.Verb) {
		if _, ok := Path(req.Args, k); !ok {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		return ExecuteResult{
			Status:        ExecFailed,
			Errors:        []string{"settler.invalid_args"},
			MissingFields: missing,
		}
	}
	if field := validateAmount(req.Args); field != "" {
		return ExecuteResult{
			Status:        ExecFailed,
			Errors:        []string{"settler.invalid_args"},
			MissingFields: []string{field},
		}
	}
	if !railSupported(req.Rail) {
		return ExecuteResult{Status: ExecFailed, Errors: []string{"settler.unsupported_rail"}}
	}

	idemKey := req.IdempotencyKey
	if idemKey == "" {
		var err error
		idemKey, err = DeriveIdempotencyKey(req.CommitmentCid, req.Verb, req.Rail, req.Args, 0)
		if err != nil {
			return ExecuteResult{Status: ExecFailed, Errors: []string{"settler.invalid_args"}}
		}
	}

	if store != nil {
		if cached, ok := store[idemKey]; ok {
			return ExecuteResult{Status: ExecSuccess, Receipt: cached}
		}
	}

	ah, err := ArgsHash(req.Args)
	if err != nil {
		return ExecuteResult{Status: ExecFailed, Errors: []string{"settler.invalid_args"}}
	}

	requestID := req.RequestID
	if requestID == "" {
		requestID = "exec-" + truncate(idemKey, 12)
	}

	anchor := NewObject()
	anchor.Set("target", anchorTarget)
	anchor.Set("locator", "batch-"+truncate(idemKey[1:], 9))
	anchor.Set("status", "pending")

	body := NewObject()
	body.Set("request_id", requestID)
	body.Set("commitment_cid", req.CommitmentCid)
	body.Set("verb", req.Verb)
	body.Set("rail", req.Rail)
	body.Set("args_hash", ah)
	body.Set("idempotency_key", idemKey)
	body.Set("status", "success")
	body.Set("locator", "MOCK-"+truncate(idemKey[1:], 17))
	body.Set("anchors", []Value{anchor})
	body.Set("ts", ts)

	canonical, err := Canonicalize(body)
	if err != nil {
		return ExecuteResult{Status: ExecFailed, Errors: []string{"settler.invalid_args"}}
	}
	xr := body.Clone()
	xr.Set("cid", CidFromCanonical(canonical))
	xr.Set("sig", SignDetached([]byte(canonical), priv, kid))

	if store != nil {
		store[idemKey] = xr
	}
	return ExecuteResult{Status: ExecSuccess, Receipt: xr}
}
